src/IP.php

```diff
@@ -69,6 +74,121 @@ abstract class IP
      */
     protected $class;
 
+    /**
+     * Constructor tries to guess what is the $ip.
+     *
+     * @param $ip mixed String, binary string, int or float
+     */
+    public function __construct($ip)
+    {
+        if (is_int($ip)) {
+            $this->fromInt($ip);
+
+            return;
+        }
+
+        // float (or double) with an integer value
+        if (is_float($ip) && $ip == floor($ip)) {
+            $this->fromFloat($ip);
+
+            return;
+        }
+
+        if (is_string($ip)) {
+            $this->fromString($ip);
+
+            return;
+        }
+
+        if ((is_resource($ip) && get_resource_type($ip) === 'GMP integer') || $ip instanceof \GMP) {
+            $this->fromGMP($ip);
+
+            return;
+        }
+
+        throw new \InvalidArgumentException(sprintf('Unsupported argument type: "%s".', gettype($ip)));
+    }
+
+    /**
+     * @param int $ip
+     */
+    private function fromInt($ip)
```

Comment:
For the from* methods, it would be nice to have type hinting. As discussed in #8 it makes sense to only support PHP >= 7.

I also think these from* should be static and return $ip.

Reply:
When you say "should return `$ip`", do you mean return the `\GMP` instance or an instance of `PhpIP\IP`?

Reply:
I mean the GMP instance, so in the constructor it would look like "$this->ip = self::fromInt($ip)".